tests/mir-opt/lower_intrinsics.transmute_uninhabited.LowerIntrinsics.diff

```diff
@@ -0,0 +1,22 @@
+- // MIR for `transmute_uninhabited` before LowerIntrinsics
++ // MIR for `transmute_uninhabited` after LowerIntrinsics
+  
+  fn transmute_uninhabited(_1: ()) -> Never {
+      debug u => _1;                       // in scope 0 at $DIR/lower_intrinsics.rs:+0:37: +0:38
+      let mut _0: Never;                   // return place in scope 0 at $DIR/lower_intrinsics.rs:+0:47: +0:52
+      let mut _2: ();                      // in scope 0 at $DIR/lower_intrinsics.rs:+1:47: +1:48
+      scope 1 {
+      }
+  
+      bb0: {
+          StorageLive(_2);                 // scope 1 at $DIR/lower_intrinsics.rs:+1:47: +1:48
+          _2 = _1;                         // scope 1 at $DIR/lower_intrinsics.rs:+1:47: +1:48
+-         _0 = transmute::<(), Never>(move _2); // scope 1 at $DIR/lower_intrinsics.rs:+1:14: +1:49
+-                                          // mir::Constant
+-                                          // + span: $DIR/lower_intrinsics.rs:48:14: 48:46
+-                                          // + literal: Const { ty: unsafe extern "rust-intrinsic" fn(()) -> Never {transmute::<(), Never>}, val: Value(<ZST>) }
++         _0 = move _2 as Never (Transmute); // scope 1 at $DIR/lower_intrinsics.rs:+1:14: +1:49
```

Comment:
I don't think this is right. Previously you did get a panic/abort, but now you get UB before hitting the `unreachable`, right?

Reply:
we still generate an abort (at least for ssa/LLVM) at the cast site. It was always UB before, too afaict.

Reply:
Yeah, previously we had a function call without return block, which is UB if it ever returns.

Reply:
The transmute intrinsic previously panicked. cg_clif doesn't abort on a write_cvalue_maybe_transmute of an unreachable type.

Reply:
> The transmute intrinsic previously panicked. 

Looking at the [diff](https://github.com/rust-lang/rust/pull/108442/files#diff-6eb34d22bb08ffb647a2da4423957658ad0be3f51d9991a172fe5525b36cd8d8) in this PR it seems to have created an `unreachable` LLVM terminator, i.e., UB?

Reply:
Right, cg_clif does panic however. Probably should have been an abort instead.

https://github.com/rust-lang/rust/blob/ef03fda339923e659d3d3ca3321de887316d2807/compiler/rustc_codegen_cranelift/src/intrinsics/mod.rs#L563-L566

Reply:
Oh, I lost that `codegen_panic` in this change 😬

@bjorn3 it sounds like I should have cg_clif [`trap`](https://docs.rs/cranelift-codegen/latest/cranelift_codegen/ir/trait.InstBuilder.html#method.trap) in the places where I have LLVM [`llvm.trap`](https://llvm.org/docs/LangRef.html#llvm-trap-intrinsic)ing?

Reply:
Actually, things might be fine as-is.  A call to `mem::transmute` that creates an uninhabited type [becomes](https://github.com/rust-lang/rust/pull/108442/files#diff-a092d178a9a0e9314a01f2e7cfe730939d540866cdffc98d5a893eabfa78ff71R19) a `CastKind::Transmute` followed by `TerminatorKind::Unreachable`, and that `Unreachable` will trap in cranelift:

https://github.com/rust-lang/rust/blob/1db9c061d387a66ab16a90ec5a1b90adf216e2da/compiler/rustc_codegen_cranelift/src/base.rs#L486-L488